include/irr/asset/CTexturePacker.h

```diff
@@ -163,83 +133,144 @@ class ITexturePacker
 
 class ICPUTexturePacker : protected ITexturePacker
 {
-    uint32_t megaimgLayerSize(uint32_t _pgSz, uint32_t _tilesPerDim) const
+    uint32_t physAddrTexLayerSz(uint32_t _pgSz, uint32_t _tilesPerDim) const
     {
         return _tilesPerDim*(TILE_PADDING + _pgSz) + TILE_PADDING;
     }
-    core::vector2du32_SIMD pageCoords(uint32_t _x, uint32_t _y, uint32_t _pgSz) const
+    core::vector3du32_SIMD pageCoords(STexOffset _txoffset, uint32_t _pgSz) const
     {
-        core::vector2du32_SIMD coords(_x,_y);
+        core::vector3du32_SIMD coords(_txoffset.x(),_txoffset.y(), 0u);
         coords *= (_pgSz + TILE_PADDING);
         coords += TILE_PADDING;
+        coords.z = _txoffset.layer();
         return coords;
     }
 
 public:
-    ICPUTexturePacker(E_FORMAT _format, uint32_t _pgSzxy = 256u, uint32_t _tilesPerDim = 32u, uint32_t _numLayers = 4u) :
-        ITexturePacker(_pgSzxy, _tilesPerDim, _numLayers)
+    ICPUTexturePacker(E_FORMAT _format, uint32_t _pgSzxy = 256u, uint32_t _tilesPerDim = 32u, uint32_t _numLayers = 4u, uint32_t _tilePad = 9u/*max_aniso/2+1*/) :
+        ITexturePacker(_pgSzxy, _tilesPerDim, _numLayers), TILE_PADDING(_tilePad)
     {
-        assert(core::isPoT(_tilesPerDim));
+        {
+            const uint32_t SZ = physAddrTexLayerSz(_pgSzxy, _tilesPerDim);
+
+            ICPUImage::SCreationParams params;
+            params.extent = { SZ,SZ,1u };
+            params.format = _format;
+            params.arrayLayers = _numLayers;
+            params.mipLevels = 1u;
+            params.type = IImage::ET_2D;
+            params.samples = IImage::ESCF_1_BIT;
+            params.flags = static_cast<IImage::E_CREATE_FLAGS>(0);
+
+            m_physAddrTex = ICPUImage::create(std::move(params));
+            auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<IImage::SBufferCopy>>(1ull);
+            auto& region = regions->front();
+            region.imageSubresource.mipLevel = 0u;
+            region.imageSubresource.baseArrayLayer = 0u;
+            region.imageSubresource.layerCount = _numLayers;
+            region.bufferOffset = 0u;
+            region.bufferRowLength = SZ;
+            region.bufferImageHeight = 0u; //tightly packed
+            region.imageOffset = { 0u, 0u, 0u };
+            region.imageExtent = { SZ,SZ,1u };
+            auto buffer = core::make_smart_refctd_ptr<ICPUBuffer>(getTexelOrBlockBytesize(_format) * SZ * SZ * _numLayers);
+            m_physAddrTex->setBufferAndRegions(std::move(buffer), regions);
+        }
+        {
+            ICPUImage::SCreationParams params;
+            params.arrayLayers = _numLayers;
```

Comment:
the virtual address range can have a completely independent number of layers, and also there's no reason to have more than 1 (because even a 16kx16k texture will give you crazy virtual texcoord range)